ARROW-11981: [C++] Implement Union ExecNode

aocsa opened the pull request:
UnionDataset allows Fragments of multiple schemas and differing file formats to be scanned together as a single Dataset. This is useful functionality but makes the Dataset interface somewhat difficult to reason about since it must be general enough to accommodate UnionDataset.

After ARROW-11928 it will probably be more natural to support unioning of datasets through a subclass of ExecNode. Reconciliation of differing schemas can then be trivially handled by a full ProjectNode.

Note this would obviate both ARROW-11001 and ARROW-11749. In addition, Dataset could be simplified to a concrete class containing a set of compatibly typed/formatted Fragments.

lidavidm commented on cpp/src/arrow/compute/exec/union_node.cc:
```diff
@@ -0,0 +1,131 @@
+namespace compute {
+
+struct UnionNode : ExecNode {
+  UnionNode(ExecNode* lhs_input, ExecNode* rhs_input, ExecContext* ctx)
```
Most other nodes take ExecPlan* plan as the first input now. 

Additionally, is there any reason why this couldn't be made a variadic node?

nirandaperera replied:
can we have variable number of inputs in the exec node registry ATM? :thinking:

lidavidm replied:
Everything takes vector<ExecNode*>: https://github.com/apache/arrow/blob/5c5a0d63a42dc8d5ecab5996574c466f2e9c2ed5/cpp/src/arrow/compute/exec/exec_plan.h#L239-L276

lidavidm commented on cpp/src/arrow/compute/exec/union_node.cc:
```diff
@@ -0,0 +1,131 @@
+
+ private:
+  ExecContext* ctx_;
+  std::mutex mutex_;
+  int batch_count_{0};
+  int input_count_{0};
```
There's a helper called AtomicCounter to manage this sort of thing without having to keep the mutex.

lidavidm commented on cpp/src/arrow/compute/exec/union_node.cc:
```diff
@@ -0,0 +1,131 @@
+    std::unique_lock<std::mutex> lk(mutex_);
+    if (input_count_ == 2) {
+      finished_.MarkFinished();
+      outputs_[0]->InputFinished(this, batch_count_);
```
InputFinished isn't guaranteed to be called after calls to InputReceived finish. Instead we should add `num_total` to a counter here, then when we've gotten InputFinished from all inputs, we can then call InputFinished on the output.

We shouldn't MarkFinished until we get all the batches. 

You can see this pattern in aggregate_node.cc: https://github.com/apache/arrow/blob/5c5a0d63a42dc8d5ecab5996574c466f2e9c2ed5/cpp/src/arrow/compute/exec/aggregate_node.cc#L483-L492 and https://github.com/apache/arrow/blob/5c5a0d63a42dc8d5ecab5996574c466f2e9c2ed5/cpp/src/arrow/compute/exec/aggregate_node.cc#L464-L475

lidavidm replied:
So it would look something like (assuming we use AtomicCounter)

```cpp
input_count_.SetTotal(2); // or inputs.size() if we want to make this a variadic node
// ...

// InputFinished
total_batches_ += num_total;
if (input_count_.Increment()) {
  outputs_[0]->InputFinished(this, total_batches_);
  if (batch_count.SetTotal(total_batches_)) {
    finished_.MarkFinished();
  }
}

// InputReceived
outputs_[0]->InputReceived(this, seq, std::move(batch));
if (batch_count_.Increment()) {
  finished_.MarkFinished();
}

// StopProducing
if (batch_count_.Cancel()) finished_.MarkFinished();
```

lidavidm replied:
I wonder if we could make some sort of testing node that purposely permuted the order of operations to try to catch this…

aocsa replied:
Thanks I followed your suggest to update this PR

lidavidm commented on cpp/src/arrow/compute/exec/union_node_test.cc:
```diff
@@ -0,0 +1,188 @@
+void CheckRunOutput(const BatchesWithSchema& l_batches,
+                    const BatchesWithSchema& r_batches,
+                    const BatchesWithSchema& exp_batches, bool parallel = false) {
+  SCOPED_TRACE("serial");
```
This should depend on `parallel`?

lidavidm commented on cpp/src/arrow/compute/exec/union_node_test.cc:
```diff
@@ -0,0 +1,188 @@
+
+  ASSERT_OK_AND_ASSIGN(auto plan, ExecPlan::Make());
+
+  ExecNodeOptions Union_options{};
+  Declaration union_decl{"union", Union_options};
```
```suggestion
  ExecNodeOptions union_options{};
  Declaration union_decl{"union", union_options};
```

lidavidm replied:
(You should be able to get away with `Declaration union_decl{"union", {}}`?)

lidavidm commented on cpp/src/arrow/compute/exec/union_node_test.cc:
```diff
@@ -0,0 +1,188 @@
+
+TEST_P(UnionTest, TestNonEmpty) { RunNonEmptyTest(std::get<0>(GetParam())); }
+
+TEST_P(UnionTest, TestEmpty) { RunEmptyTest(std::get<0>(GetParam())); }
```
Why not just inline these?

aocsa replied:
Sure

lidavidm commented on cpp/src/arrow/compute/exec/union_node_test.cc:
```diff
@@ -0,0 +1,188 @@
+                .AddToPlan(plan.get()));
+
+  ASSERT_FINISHES_OK_AND_ASSIGN(auto res, StartAndCollect(plan.get(), sink_gen));
+}
```
You should be able to check the batches here too.

lidavidm reviewed (commented):
Thanks for doing this. I think we can clean up the implementation here using AtomicCounter instead of manually synchronizing with mutexes.

nirandaperera commented on cpp/src/arrow/compute/exec/union_node.cc:
```diff
@@ -0,0 +1,131 @@
+      return;
+    }
+    {
+      std::unique_lock<std::mutex> lock(mutex_);
```
I think it would be better to use AtomicCounter here 
https://github.com/apache/arrow/blob/d943b8bf22149775c84b5afbda5c34cb9a19d93c/cpp/src/arrow/compute/exec/util.h#L194

nirandaperera commented on cpp/src/arrow/compute/exec/union_node.cc:
```diff
@@ -0,0 +1,131 @@
+  Future<> finished_ = Future<>::MakeFinished();
+};
+
+Result<ExecNode*> MakeUnionNode(ExecNode* lhs_input, ExecNode* rhs_input) {
```
I think we need to validate the lhs and rhs schemas here. They should be the same.

nirandaperera reviewed (changes requested):
left some comments. :-) Good work @aocsa

aocsa commented:
Thanks for the review @lidavidm, @nirandaperera. 
I update this PR with the following main changes: 
 
1. The use AtomicCounter
2. Handle variable number of inputs 
3. Validate input schemas

lidavidm commented on cpp/src/arrow/compute/exec/union_node.cc:
```diff
@@ -0,0 +1,150 @@
+namespace compute {
+
+namespace {
+std::vector<std::string> GetInputLabels(ExecNode::NodeVector inputs) {
```
```suggestion
std::vector<std::string> GetInputLabels(const ExecNode::NodeVector& inputs) {
```

lidavidm commented on cpp/src/arrow/compute/exec/union_node.cc:
```diff
@@ -0,0 +1,150 @@
+                 /*output_schema=*/inputs[0]->output_schema(),
+                 /*num_outputs=*/1) {
+    if (this->input_count_.SetTotal(static_cast<int>(inputs.size()))) {
+      finished_.MarkFinished();
```
Hmm, this branch shouldn't be possible unless inputs == 0 which isn't allowed by Make.

lidavidm commented on cpp/src/arrow/compute/exec/union_node.cc:
```diff
@@ -0,0 +1,150 @@
+    for (auto input : inputs) {
+      if (!input->output_schema()->Equals(schema)) {
+        return Status::Invalid(
+            "Constructing a `UnionNode` with inputs with different schemas");
```
Nit: it might be nice to provide the schemas to aid debugging, something like

```cpp
return Status::Invalid("UnionNode input schemas must all match, first schema was: ", *schema, " got schema: ", *input->output_schema);
```

lidavidm commented on cpp/src/arrow/compute/exec/union_node.cc:
```diff
@@ -0,0 +1,150 @@
+    return plan->EmplaceNode<UnionNode>(plan, std::move(inputs));
+  }
+
+  inline bool IsLeftInput(ExecNode* input) { return input == inputs_[0]; }
```
This looks unused now.

lidavidm commented on cpp/src/arrow/compute/exec/union_node.cc:
```diff
@@ -0,0 +1,150 @@
+    if (finished_.is_finished()) {
+      return;
+    }
+    outputs_[0]->InputReceived(this, seq, std::move(batch));
```
I wonder if it actually makes sense to forward seq like this. As I understand nodes can't really rely on this anyways so I don't think it harms things, but it seems odd to even have the parameter in that case. CC @bkietz for opinions. 

(We could modify `Increment` to also give the count, if we wanted to give a unique seq_num to each batch we emit here.)

bkietz replied:
I think `seq` should be ultimately removed. It's a confusing parameter which we don't provide guarantees on. I'll add a JIRA for this anon

lidavidm replied:
Thanks. @aocsa we can leave this as-is then.

bkietz replied:
https://issues.apache.org/jira/browse/ARROW-13660

lidavidm commented on cpp/src/arrow/compute/exec/union_node.cc:
```diff
@@ -0,0 +1,150 @@
+    return Status::OK();
+  }
+
+  void PauseProducing(ExecNode* output) override {}
+
+  void ResumeProducing(ExecNode* output) override {}
```
I think we can trivially forward pause/resume to our inputs.

lidavidm replied:
Or if not, we should make a followup JIRA.

bkietz replied:
https://issues.apache.org/jira/browse/ARROW-13659

lidavidm commented on cpp/src/arrow/compute/exec/union_node.cc:
```diff
@@ -0,0 +1,150 @@
+  void StopProducing(ExecNode* output) override {
+    DCHECK_EQ(output, outputs_[0]);
+    if (batch_count_.Cancel()) {
+      finished_.MakeFinished();
```
```suggestion
      finished_.MarkFinished();
```

lidavidm commented on cpp/src/arrow/compute/exec/union_node.cc:
```diff
@@ -0,0 +1,150 @@
+      finished_.MakeFinished();
+    }
+    inputs_[0]->StopProducing(this);
+  }
```
This should call StopProducing on every input.

lidavidm commented on cpp/src/arrow/compute/exec/union_node_test.cc:
```diff
@@ -0,0 +1,203 @@
+  for (bool parallel : {false, true}) {
+    RunEmptyTest(parallel);
+  }
+}
```
I think it'd be good to add a test that tries to create a union node with 0 inputs, and with differing schemas, to ensure those cases are covered.

lidavidm commented on cpp/src/arrow/compute/exec/union_node_test.cc:
```diff
@@ -0,0 +1,203 @@
+                                                                  /*slow=*/false)}});
+  // add right source
+  Union.inputs.emplace_back(Declaration{
+      "source", SourceNodeOptions{r_batches.schema, r_batches.gen(parallel,
```
Maybe we could make this test use three inputs? We could even just duplicate one of the existing inputs.

lidavidm commented on cpp/src/arrow/compute/exec/union_node.cc:
```diff
@@ -0,0 +1,150 @@
+  static Result<ExecNode*> Make(ExecPlan* plan, std::vector<ExecNode*> inputs,
+                                const ExecNodeOptions& options) {
+    RETURN_NOT_OK(ValidateExecNodeInputs(plan, inputs, inputs.size(), "UnionNode"));
+    if (inputs.size() < 2) {
```
It's kinda pointless, but allowing one input would be a little more consistent. (It might make life a little easier to not have to special case the n = 1 case when generating exec plans.)

lidavidm reviewed (commented):
Thanks for the changes! This is much cleaner now.

aocsa commented:
Thanks @lidavidm, I updated this PR addressing feedback and ensuring the following cases are covered.
- union node with variable number of inputs with different schemas 
- union node with 0 inputs
- union node with inputs with empty batches

lidavidm commented on cpp/src/arrow/compute/exec/union_node.cc:
```diff
@@ -48,31 +48,29 @@ struct UnionNode : ExecNode {
-    if (this->input_count_.SetTotal(static_cast<int>(inputs.size()))) {
-      finished_.MarkFinished();
-    }
+    ARROW_DCHECK(this->input_count_.SetTotal(static_cast<int>(inputs.size())) == false);
```
This won't work - this will get compiled out in release mode.

lidavidm replied:
In particular I think this is responsible for the MinGW CI failures (which compile in release mode vs the others which compile in debug mode)

lidavidm commented on cpp/src/arrow/compute/exec/union_node_test.cc:
```diff
@@ -17,187 +17,126 @@
-  TestUnionRandom(std::get<0>(GetParam()), std::get<1>(GetParam()), kNumBatches,
-                  kBatchSize);
-}
+TEST_F(TestUnionNode, TestEmpty) { this->CheckUnionExecNode(0, 0, false); }
```
Nit: usually for constant parameters like this, we prefix the value with the argument name. See https://google.github.io/styleguide/cppguide.html#Function_Argument_Comments

lidavidm replied:
```suggestion
TEST_F(TestUnionNode, TestEmpty) {
  this->CheckUnionExecNode(/*num_input_nodes*/0, /*num_batches=*/0, /*parallel=*/false);
}
```

lidavidm replied:
And conversely above in TestNonEmpty, the comment is redundant since there's a constant.

lidavidm commented on cpp/src/arrow/compute/exec/union_node_test.cc:
```diff
@@ -17,187 +17,126 @@
+  ::arrow::random::RandomArrayGenerator rng_;
 
-  auto actual = StartAndCollect(plan.get(), sink_gen);
+  static constexpr int kNumBatches = 10;
```
nit: put constants towards the top. See https://google.github.io/styleguide/cppguide.html#Declaration_Order

lidavidm reviewed (commented):
Thanks, just a few final comments. 

Also there are a few lint failures: https://github.com/apache/arrow/pull/10927/checks?check_run_id=3350092948#step:7:4091

You can use docker-compose to lint code: https://arrow.apache.org/docs/developers/cpp/development.html#code-style-linting-and-ci

nirandaperera commented on cpp/src/arrow/compute/exec/union_node.cc:
```diff
@@ -0,0 +1,150 @@
+  }
+
+  void InputFinished(ExecNode* input, int num_total) override {
+    ARROW_DCHECK(std::find_if(inputs_.begin(), inputs_.end(), [input](ExecNode* n) {
+                   return n == input;
+                 }) != inputs_.end());
```
```suggestion
    ARROW_DCHECK(std::find(inputs_.begin(), inputs_.end(), input) != inputs_.end());
```
There's no need to use `find_if` here IMO.

nirandaperera commented on cpp/src/arrow/compute/exec/union_node.cc:
```diff
@@ -0,0 +1,150 @@
+    for (auto input : inputs) {
+      if (!input->output_schema()->Equals(schema)) {
+        return Status::Invalid(
+            "UnionNode input schemas must all match, first schema was: ", *schema,
+            " got schema: ", *input->output_schema());
```
```suggestion
            "UnionNode input schemas must all match, first schema was: ", *schema.ToString(),
            " got schema: ", *input->output_schema().ToString());
```

nirandaperera commented on cpp/src/arrow/compute/exec/union_node.cc:
```diff
@@ -0,0 +1,150 @@
+                   return n == input;
+                 }) != inputs_.end());
+
+    total_batches_.fetch_add(num_total);
+
+    if (input_count_.Increment()) {
+      outputs_[0]->InputFinished(this, total_batches_.load());
+      if (batch_count_.SetTotal(total_batches_.load())) {
+        finished_.MarkFinished();
+      }
+    }
```
I feel like this `total_batches_` is redundant. Cant we do this instead?
```suggestion
    batch_count_.Increment(num_total);

    if (input_count_.Increment()) {
      outputs_[0]->InputFinished(this, total_batches_.load());
      int total_batches = batch_count_.count();
      if (batch_count_.SetTotal(total_batches)) {
        finished_.MarkFinished();
      }
    }
```

aocsa replied:
you mean:
```cpp
if (input_count_.Increment()) {
      int total_batches = batch_count_.count();
      outputs_[0]->InputFinished(this, total_batches);
      if (batch_count_.SetTotal(total_batches)) {
        finished_.MarkFinished();
      }
    }
```
right?

nirandaperera replied:
Please ignore this suggestion. This leads to a logic error as pointed out by @lidavidm

lidavidm commented on cpp/src/arrow/compute/exec/union_node.cc:
```diff
@@ -0,0 +1,145 @@
+      if (batch_count_.SetTotal(total_batches)) {
+        finished_.MarkFinished();
+      }
+    }
```
This has a logic error again: InputFinished can be called before InputReceived. We need to make use of num_total here.

lidavidm replied:
@nirandaperera I don't think the proposal in https://github.com/apache/arrow/pull/10927#discussion_r690444907 works because we need to track both 1) batches seen so far and 2) total batches as reported by our inputs and we can't reuse the same counter for this, since setting the total of an AtomicCounter is effectively a one-time-only operation. (You could imagine ways to stuff it into the encoding: for example, an AccumulateTotal() that adds a *negative* value to total_, and a separate method SetTotal(void) that inverts the value - that would work modulo having to express all this with just atomics.)

nirandaperera replied:
Ah. thanks @lidavidm for pointing that out. this certainly is a logic error. In that case, the previous `atomic<int>` approach is the way to go! @aocsa sorry for the confusion.

aocsa replied:
got it, thanks @lidavidm

lidavidm commented:
Merged. Thanks!

lidavidm commented:
I filed ARROW-13653 as a follow-up.

aocsa commented:
Thanks @lidavidm :+1:
